File: javascript/workflows/batch-pronunciation-folder.js
/**
 * Batch Pronunciation Assessment — Vocametrix API workflow (Node.js)
 *
 * Walks a folder of WAV files, runs pronunciation assessment on each,
 * writes results to a CSV. Uses exponential backoff for 429 rate limits.
 *
 * Usage:
 *   node batch-pronunciation-folder.js /path/to/wav/folder
 *   node batch-pronunciation-folder.js ./recordings --text "Say ah" --locale fr-FR --output results.csv --workers 4
 */

import fs from 'fs';
import path from 'path';
import { createWriteStream } from 'fs';
import FormData from 'form-data';
import 'dotenv/config';

const API_KEY = process.env.VOCAMETRIX_API_KEY;
const BASE_URL = 'https://platform.vocametrix.com';
const HEADERS = { 'X-API-Key': API_KEY, 'Content-Type': 'application/json' };

function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

async function uploadViaBlobUrl(audioPath) {
  const res = await fetch(`${BASE_URL}/api/get-blob-url`, { headers: HEADERS });
  if (!res.ok) throw new Error(`get-blob-url failed: ${await res.text()}`);
  const { uploadURL, blobURL } = await res.json();
  const put = await fetch(uploadURL, {
    method: 'PUT',
    body: fs.readFileSync(audioPath),
    headers: { 'x-ms-blob-type': 'BlockBlob', 'Content-Type': 'audio/wav' },
  });
  if (!put.ok) throw new Error(`Azure upload failed: ${await put.text()}`);
  return blobURL;
}

async function assessOne(audioPath, text, locale, maxRetries = 3) {
  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    const blobURL = await uploadViaBlobUrl(audioPath);
    const res = await fetch(`${BASE_URL}/api/pronunciation-assessment`, {
      method: 'POST',
      headers: HEADERS,
      body: JSON.stringify({ blobURL, referenceText: text, locale }),
    });
    if (res.status === 429) {
      const wait = (2 ** attempt) * 60_000;
      console.log(`  [${path.basename(audioPath)}] rate limited, waiting ${wait / 1000}s`);
      await sleep(wait);
      continue;
    }
    if (!res.ok) throw new Error(`pronunciation-assessment failed: ${await res.text()}`);
    return res.json();
  }
  throw new Error('Max retries exceeded');
}

async function processFolder(folder, text, locale, outputCsv, concurrency) {
  const files = fs.readdirSync(folder)
    .filter(f => f.toLowerCase().endsWith('.wav'))
    .sort()
    .map(f => path.join(folder, f));

  if (files.length === 0) { console.error('No WAV files found'); process.exit(1); }
  console.log(`Found ${files.length} files. Processing with concurrency=${concurrency}...`);

  const results = [];
  const errors = [];

  // Process in batches of `concurrency`
  for (let i = 0; i < files.length; i += concurrency) {
    const batch = files.slice(i, i + concurrency);
    await Promise.all(batch.map(async (fp) => {
      const name = path.basename(fp);
      try {
        const data = await assessOne(fp, text, locale);
        results.push({ file: name, accuracy: data.accuracyScore ?? '', fluency: data.fluencyScore ?? '',
                        completeness: data.completenessScore ?? '', pronunciation: data.pronunciationScore ?? '' });
        console.log(`  ✓ ${name.padEnd(40)}  accuracy=${data.accuracyScore ?? ''}`);
      } catch (e) {
        errors.push({ file: name, error: e.message });
        console.log(`  ✗ ${name}: ${e.message}`);
      }
    }));
  }

  results.sort((a, b) => a.file.localeCompare(b.file));

  const header = 'file,accuracy,fluency,completeness,pronunciation\n';
  const rows = results.map(r => `${r.file},${r.accuracy},${r.fluency},${r.completeness},${r.pronunciation}`).join('\n');
  fs.writeFileSync(outputCsv, header + rows + '\n', 'utf8');
  console.log(`\nWrote ${results.length} results to ${outputCsv}`);
  if (errors.length > 0) {
    console.log(`${errors.length} files failed:`);
    errors.forEach(e => console.log(`  ${e.file}: ${e.error}`));
  }
}

// --- CLI ---
const args = process.argv.slice(2);
if (args.length === 0) {
  console.error('Usage: node batch-pronunciation-folder.js <folder> [--text "..."] [--locale en-US] [--output results.csv] [--workers 3]');
  process.exit(1);
}
const folder = args[0];
const textIdx = args.indexOf('--text');
const localeIdx = args.indexOf('--locale');
const outputIdx = args.indexOf('--output');
const workersIdx = args.indexOf('--workers');

processFolder(
  folder,
  textIdx !== -1 ? args[textIdx + 1] : 'Hello, my name is Alex.',
  localeIdx !== -1 ? args[localeIdx + 1] : 'en-US',
  outputIdx !== -1 ? args[outputIdx + 1] : 'pronunciation_results.csv',
  workersIdx !== -1 ? parseInt(args[workersIdx + 1]) : 3,
).catch(err => { console.error(err.message); process.exit(1); });

File: python/workflows/batch_pronunciation_folder.py
"""
Batch Pronunciation Assessment — Vocametrix API workflow

Walks a folder of WAV files, runs pronunciation assessment on each,
writes results to a CSV. Uses a thread pool with exponential backoff
for rate-limit resilience (API limit: 100 req/15 min).

Usage:
    python batch_pronunciation_folder.py /path/to/wav/folder
    python batch_pronunciation_folder.py ./recordings --text "Say ah" --locale fr-FR --output results.csv --workers 4
"""

import os
import sys
import csv
import time
import glob
import argparse
import threading
import requests
from concurrent.futures import ThreadPoolExecutor, as_completed
from dotenv import load_dotenv

load_dotenv(os.path.join(os.path.dirname(__file__), '..', '..', '.env'))
API_KEY = os.environ['VOCAMETRIX_API_KEY']
BASE_URL = 'https://platform.vocametrix.com'
HEADERS = {'X-API-Key': API_KEY}

_print_lock = threading.Lock()


def log(msg: str):
    with _print_lock:
        print(msg)


def upload_via_blob_url(audio_path: str) -> str:
    r = requests.post(f'{BASE_URL}/api/get-blob-url', headers=HEADERS)
    r.raise_for_status()
    data = r.json()
    with open(audio_path, 'rb') as f:
        put = requests.put(
            data['uploadURL'],
            data=f,
            headers={'x-ms-blob-type': 'BlockBlob', 'Content-Type': 'audio/wav'},
        )
        put.raise_for_status()
    return data['blobURL']


def assess_one(audio_path: str, text: str, locale: str, max_retries: int = 3) -> dict:
    for attempt in range(max_retries + 1):
        try:
            blob_url = upload_via_blob_url(audio_path)
            r = requests.post(
                f'{BASE_URL}/api/pronunciation-assessment',
                headers=HEADERS,
                json={'blobURL': blob_url, 'referenceText': text, 'locale': locale},
            )
            if r.status_code == 429:
                wait = 2 ** attempt * 60  # exponential backoff: 60s, 120s, 240s
                log(f'  [{os.path.basename(audio_path)}] rate limited, waiting {wait}s (attempt {attempt + 1})')
                time.sleep(wait)
                continue
            r.raise_for_status()
            return r.json()
        except requests.RequestException as e:
            if attempt == max_retries:
                raise
            wait = 2 ** attempt * 5
            log(f'  [{os.path.basename(audio_path)}] error: {e}, retrying in {wait}s')
            time.sleep(wait)


def process_folder(folder: str, text: str, locale: str, output_csv: str, workers: int):
    wav_files = sorted(glob.glob(os.path.join(folder, '*.wav')) +
                       glob.glob(os.path.join(folder, '*.WAV')))
    if not wav_files:
        print(f'No WAV files found in {folder}')
        sys.exit(1)

    print(f'Found {len(wav_files)} WAV files. Processing with {workers} workers...')

    results = []
    errors = []

    with ThreadPoolExecutor(max_workers=workers) as executor:
        futures = {
            executor.submit(assess_one, fp, text, locale): fp
            for fp in wav_files
        }
        for future in as_completed(futures):
            fp = futures[future]
            name = os.path.basename(fp)
            try:
                data = future.result()
                accuracy = data.get('accuracyScore', '')
                fluency = data.get('fluencyScore', '')
                completeness = data.get('completenessScore', '')
                pron = data.get('pronunciationScore', '')
                results.append({
                    'file': name,
                    'accuracy': accuracy,
                    'fluency': fluency,
                    'completeness': completeness,
                    'pronunciation': pron,
                })
                log(f'  ✓ {name:40s}  accuracy={accuracy}')
            except Exception as e:
                errors.append({'file': name, 'error': str(e)})
                log(f'  ✗ {name}: {e}')

    # Sort by filename for reproducible output
    results.sort(key=lambda r: r['file'])

    with open(output_csv, 'w', newline='', encoding='utf-8') as f:
        writer = csv.DictWriter(f, fieldnames=['file', 'accuracy', 'fluency', 'completeness', 'pronunciation'])
        writer.writeheader()
        writer.writerows(results)

    print(f'\nWrote {len(results)} results to {output_csv}')
    if errors:
        print(f'{len(errors)} files failed:')
        for e in errors:
            print(f'  {e["file"]}: {e["error"]}')


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Batch pronunciation assessment for a folder of WAV files')
    parser.add_argument('folder', help='Path to folder containing WAV files')
    parser.add_argument('--text', default='Hello, my name is Alex.', help='Reference text for all files')
    parser.add_argument('--locale', default='en-US', help='Language locale')
    parser.add_argument('--output', default='pronunciation_results.csv', help='Output CSV path')
    parser.add_argument('--workers', type=int, default=3, help='Number of parallel workers (default 3)')
    args = parser.parse_args()

    process_folder(args.folder, args.text, args.locale, args.output, args.workers)
